interfaces/Portalicious/src/app/domains/project/project.api.service.ts

```diff
@@ -0,0 +1,87 @@
+import { Injectable, Signal } from '@angular/core';
+import { DomainApiService } from '~/domains/domain-api.service';
+import {
+  Project,
+  ProjectMetrics,
+  ProjectUser,
+  ProjectUserWithRolesLabel,
+} from '~/domains/project/project.model';
+
+const BASE_ENDPOINT = 'programs';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class ProjectApiService extends DomainApiService {
+  getProject(projectId: Signal<number | undefined>) {
+    return this.generateQueryOptions<Project>({
+      path: [BASE_ENDPOINT, projectId],
+      enabled: () => !!projectId(),
+    });
+  }
+
+  getProjectSummaryMetrics(projectId: Signal<number>) {
+    return this.generateQueryOptions<ProjectMetrics>({
+      path: [BASE_ENDPOINT, projectId, 'metrics/program-stats-summary'],
+    });
+  }
+
+  getProjectUsers(projectId: Signal<number>) {
+    return this.generateQueryOptions<
+      ProjectUser[],
+      ProjectUserWithRolesLabel[]
+    >({
+      path: [BASE_ENDPOINT, projectId, 'users'],
+      processResponse: (users) => {
+        const usersWithRolesLabel: ProjectUserWithRolesLabel[] = users.map(
+          (user) => ({
+            ...user,
+            allRolesLabel: user.roles.map((role) => role.label).join('; '),
+            lastLogin: user.lastLogin ? new Date(user.lastLogin) : undefined,
+          }),
+        );
+
+        return usersWithRolesLabel;
+      },
+    });
+  }
+
+  createProjectFromKobo({
+    token,
+    assetId,
+  }: {
+    token: string;
+    assetId: string;
+  }) {
+    return this.httpWrapperService.perform121ServiceRequest<
+      Project | undefined
+    >({
+      method: 'POST',
+      endpoint: BASE_ENDPOINT,
+      params: {
+        importFromKobo: true,
+        koboToken: token,
+        koboAssetId: assetId,
+      },
+    });
+  }
+
+  removeProjectUser(projectId: Signal<number>, userId: number) {
+    return this.httpWrapperService.perform121ServiceRequest<Project>({
+      method: 'DELETE',
+      endpoint: `${BASE_ENDPOINT}/${projectId().toString()}/users/${userId.toString()}`,
+    });
```

Comment:
We could potentially also tie the "invalidation of a related cache" at this point?

That would turn this class more into a "project-data management"-service then a "more dumb" "only talk to the API"-service....
I don't know how much that would look like a slippery slope or not though...

It would remove _some_ duplication maybe and would help developers "never forget to invalidate the relevant caches"....

Or could we formalize a rule like: "only data-manipulation, not business-logic" inside these API-Services?

Reply:
> Or could we formalize a rule like: "only data-manipulation, not business-logic" inside these API-Services?

That would be my preference, yes, which is why they are all `api.service.ts`